[JSC] Improve Object.defineProperties & Object.defineProperty perf via fast iteration in ToPropertyDescriptor

Constellation opened the pull request:
#### 2432909f3aecc7641714799b06b4f909322bd6db
<pre>
[JSC] Improve Object.defineProperties &amp; Object.defineProperty perf via fast iteration in ToPropertyDescriptor
<a href="https://bugs.webkit.org/show_bug.cgi?id=248754">https://bugs.webkit.org/show_bug.cgi?id=248754</a>
<a href="https://rdar.apple.com/103101281">rdar://103101281</a>

Reviewed by Alexey Shvayka.

This patch set up &quot;writable&quot; &quot;set&quot; &quot;get&quot; &quot;value&quot; &quot;enumerable&quot; &quot;configurable&quot; watchpoints onto Object.prototype so that
we can quickly extract property descriptor from Object.defineProperty / Object.defineProperties descriptor objects.
We also extract these properties by enumerating all properties instead of doing hash table lookup multiple times since
in almost all cases this is faster because property descriptor object should be small.

It broadly improves Object.defineProperty / Object.defineProperties perf.

                                                      ToT                     Patched

    redefine-property-accessor-dictionary        4.5682+-0.0286     ^      3.5888+-0.0287        ^ definitely 1.2729x faster
    define-properties-all-of-keys               12.0816+-0.0671     ^      8.8536+-0.0794        ^ definitely 1.3646x faster
    redefine-property-previous-attributes       41.2230+-0.5076     ^     29.0349+-0.2499        ^ definitely 1.4198x faster
    redefine-property-data-dictionary            3.8830+-0.0141     ^      2.9631+-0.0237        ^ definitely 1.3104x faster
    define-property-simple                       1.3220+-0.0180     ^      1.1180+-0.0131        ^ definitely 1.1825x faster
    redefine-property-data                       3.4071+-0.0218     ^      2.5390+-0.0303        ^ definitely 1.3419x faster
    redefine-property-accessor                   3.9898+-0.0219     ^      2.9743+-0.0300        ^ definitely 1.3414x faster
    define-properties-simple                     9.9033+-0.0381     ^      6.8733+-0.0296        ^ definitely 1.4408x faster

* JSTests/microbenchmarks/define-properties-all-of-keys.js: Added.
(const.properties.destroyed.set if):
(const.properties.writable.set if):
(test.Hey):
(test.):
(test):
* JSTests/microbenchmarks/define-properties-simple.js: Added.
(const.properties.destroyed.set if):
(const.properties.writable.set if):
(test.):
(test):
* JSTests/microbenchmarks/define-property-simple.js: Added.
(const.properties.destroyed.set if):
(const.properties.writable.set if):
(test.):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::tryInstallPropertyDescriptorFastPathWatchpoint):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::propertyDescriptorFastPathWatchpointSet):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::getOwnPropertyIfPropertyExists):
* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::toPropertyDescriptor):
(JSC::definePropertiesSlow):
(JSC::defineProperties):

Canonical link: <a href="https://commits.webkit.org/271972@main">https://commits.webkit.org/271972@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8739b7dc2d12ed1f005c91763afbe04b6b81758c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30221 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8890 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32726 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27338 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11090 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6126 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27324 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30525 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7466 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/26247 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6389 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6537 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/26964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34066 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/25960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/27540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [💥 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27298 "An unexpected error occured. Recent messages:OS: Monterey (12.6.8), Xcode: 14.0; Skipping applying patch since patch_id isn't provided; Checked out pull request") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32718 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/30349 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6493 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4657 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30532 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8243 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36795 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7248 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7924 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3900 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7021 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shvaikalesh commented on Source/JavaScriptCore/runtime/JSObjectInlines.h:
```diff
@@ -224,6 +224,28 @@ template<typename PropertyNameType> inline JSValue JSObject::getIfPropertyExists
     return slot.getValue(globalObject, propertyName);
 }
 
+template<typename PropertyNameType> inline JSValue JSObject::getOwnPropertyIfPropertyExists(JSGlobalObject* globalObject, const PropertyNameType& propertyName)
```
This appears unused?

shvaikalesh commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -656,44 +656,106 @@ bool toPropertyDescriptor(JSGlobalObject* globalObject, JSValue in, PropertyDesc
-            throwTypeError(globalObject, scope, "Getter must be a function."_s);
-            return false;
+    bool canUseFastPath = false;
+    if (globalObject->propertyDescriptorFastPathWatchpointSet().isStillValid() && globalObject->objectPrototypeChainIsSane() && description->inherits<JSFinalObject>() && description->getPrototypeDirect() == globalObject->objectPrototype() && description->structure()->canPerformFastPropertyEnumeration() && !hasIndexedProperties(description->indexingType())) {
```
Do we actually care about `!hasIndexedProperties` here?

shvaikalesh commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -656,44 +656,106 @@ bool toPropertyDescriptor(JSGlobalObject* globalObject, JSValue in, PropertyDesc
-            return false;
+    bool canUseFastPath = false;
+    if (globalObject->propertyDescriptorFastPathWatchpointSet().isStillValid() && globalObject->objectPrototypeChainIsSane() && description->inherits<JSFinalObject>() && description->getPrototypeDirect() == globalObject->objectPrototype() && description->structure()->canPerformFastPropertyEnumeration() && !hasIndexedProperties(description->indexingType())) {
+        if (!description->staticPropertiesReified()) {
```
nit: It's extremely unlikely for a final object to have non-reified static properties. If we remove this condition, `fastForEachPropertyWithSideEffectFreeFunctor` would correctly handle such an unlikely case by returning `false`.

Constellation committed:
[JSC] Improve Object.defineProperties & Object.defineProperty perf via fast iteration in ToPropertyDescriptor
https://bugs.webkit.org/show_bug.cgi?id=248754
rdar://103101281

Reviewed by Alexey Shvayka.

This patch set up "writable" "set" "get" "value" "enumerable" "configurable" watchpoints onto Object.prototype so that
we can quickly extract property descriptor from Object.defineProperty / Object.defineProperties descriptor objects.
We also extract these properties by enumerating all properties instead of doing hash table lookup multiple times since
in almost all cases this is faster because property descriptor object should be small.

It broadly improves Object.defineProperty / Object.defineProperties perf.

                                                      ToT                     Patched

    redefine-property-accessor-dictionary        4.5682+-0.0286     ^      3.5888+-0.0287        ^ definitely 1.2729x faster
    define-properties-all-of-keys               12.0816+-0.0671     ^      8.8536+-0.0794        ^ definitely 1.3646x faster
    redefine-property-previous-attributes       41.2230+-0.5076     ^     29.0349+-0.2499        ^ definitely 1.4198x faster
    redefine-property-data-dictionary            3.8830+-0.0141     ^      2.9631+-0.0237        ^ definitely 1.3104x faster
    define-property-simple                       1.3220+-0.0180     ^      1.1180+-0.0131        ^ definitely 1.1825x faster
    redefine-property-data                       3.4071+-0.0218     ^      2.5390+-0.0303        ^ definitely 1.3419x faster
    redefine-property-accessor                   3.9898+-0.0219     ^      2.9743+-0.0300        ^ definitely 1.3414x faster
    define-properties-simple                     9.9033+-0.0381     ^      6.8733+-0.0296        ^ definitely 1.4408x faster

* JSTests/microbenchmarks/define-properties-all-of-keys.js: Added.
(const.properties.destroyed.set if):
(const.properties.writable.set if):
(test.Hey):
(test.):
(test):
* JSTests/microbenchmarks/define-properties-simple.js: Added.
(const.properties.destroyed.set if):
(const.properties.writable.set if):
(test.):
(test):
* JSTests/microbenchmarks/define-property-simple.js: Added.
(const.properties.destroyed.set if):
(const.properties.writable.set if):
(test.):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::tryInstallPropertyDescriptorFastPathWatchpoint):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::propertyDescriptorFastPathWatchpointSet):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::getOwnPropertyIfPropertyExists):
* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::toPropertyDescriptor):
(JSC::definePropertiesSlow):
(JSC::defineProperties):

Canonical link: https://commits.webkit.org/271972@main
